<!DOCTYPE html>
<html>
<head>
</head>
<body>
<h1 id="cutruntools">CutRunTools</h1>
<p>This package contains the pipeline for conducting a CUT&amp;RUN analysis.
The pipeline comprises of read trimming, alignment steps, motif finding steps, and finally the motif footprinting step.</p>
<p>To get started, please see INSTALL.md about how to set up the pipeline.</p>
<p>Once the package is installed, please see USAGE.md to use it (but basic usage is provided below).</p>
<h2 id="basic-usage">Basic Usage</h2>
<p>CutRunTools requires a JSON configuration file which specifies all that is needed to run an analysis.
A sample configuration file is below. Note the important settings in bold.</p>
<pre><code>
        {
                "Rscriptbin": "/n/app/R/3.3.3/bin",
                "pythonbin": "/n/app/python/2.7.12/bin",
                "trimmomaticbin": "/n/app/trimmomatic/0.36/bin",
                "trimmomaticjarfile": "trimmomatic-0.36.jar",
                "bowtie2bin": "/n/app/bowtie2/2.2.9/bin",
                "samtoolsbin": "/n/app/samtools/1.3.1/bin",
                "adapterpath": "/home/qz64",
                "picardbin": "/n/app/picard/2.8.0/bin",
                "picardjarfile": "picard-2.8.0.jar",
                "macs2bin": "/n/app/macs2/2.1.1.20160309/bin",
                "kseqbin": "/home/qz64",
                "memebin": "/home/qz64/meme/bin",
                "bedopsbin": "/home/qz64/bin",
                "bedtoolsbin": "/n/app/bedtools/2.26.0/bin",
                "makecutmatrixbin": "/home/qz64/.local/bin",
                <b>"bt2idx"</b>: "/n/groups/shared_databases/bowtie2_indexes",
                <b>"genome_sequence"</b>: "/home/qz64/chrom.hg19/hg19.fa",
                "extratoolsbin": "/home/qz64",
                "extrasettings": "/home/qz64",
                <b>"input/output"</b>: {
                        <b>"fastq_directory"</b>: "/n/scratch2/qz64/Nan_18_aug23/Nan_run_19",
                        <b>"workdir"</b>: "/n/scratch2/qz64/workdir",
                        <b>"fastq_sequence_length"</b>: 40,
                        <b>"organism_build"</b>: "hg19"
                },
                "motif_finding": {
                        "num_bp_from_summit": 150,
                        "num_peaks": 5000,
                        "total_peaks": 15000,
                        "motif_scanning_pval": 0.0005,
                        "num_motifs": 20
                },
                <b>"cluster"</b>: {
                        <b>"email"</b>: "bernardzhu@gmail.com",
                        "step_alignment": {
                                "queue": "short",
                                "memory": 32000,
                                "time_limit": "0-12:00"
                        },
                        "step_process_bam": {
                                "queue": "short",
                                "memory": 32000,
                                "time_limit": "0-12:00"
                        },
                        "step_motif_find": {
                                "queue": "short",
                                "memory": 32000,
                                "time_limit": "0-12:00"
                        },
                        "step_footprinting": {
                                "queue": "short",
                                "memory": 32000,
                                "time_limit": "0-12:00"
                        }
                }
        }
</code></pre>

<p><code>fastq_directory</code> is the directory containing paired-end CUT&amp;RUN sequences (with _R1_001.fastq.gz and _R2_001.fastq.gz suffix). <code>organism_build</code> is one of supported genome assemblies: hg38, hg19, mm10, and mm9. <code>adapterpath</code> contains Illumina Truseq3-PE adapter sequences (we provide them). <code>genome_sequence</code> is the whole-genome <strong>masked</strong> sequence which matches with the appropriate organism build.</p>
<h3 id="create-job-submission-scripts">Create job submission scripts</h3>
<pre><code>./create_scripts<span class="hljs-selector-class">.py</span> config.json
</code></pre><p>This creates a set of Slurm job-submission scripts based on the configuration file above (named config.json) in the <code>workdir</code> directory. The scripts can be directly, easily executed.</p>
<pre><code>./validate<span class="hljs-selector-class">.py</span> config.json
</code></pre><p>This script checks that your configuration file is correct and all paths are correct.</p>
<h3 id="four-step-process">Four-step process</h3>
<p>With the scripts created, we can next perform the analysis.</p>
<ol>
<li><p><strong>Read trimming, alignment.</strong> We suppose the <code>workdir</code> is defined as <code>/n/scratch2/qz64/workdir</code></p>
<pre><code>cd /n/scratch2/qz64/workdir
sbatch ./integrated<span class="hljs-selector-class">.sh</span> CR_BCL11A_W9_r1_S17_R1_001<span class="hljs-selector-class">.fastq</span><span class="hljs-selector-class">.gz</span>
</code></pre><p>The parameter is the fastq file. Always use the _R1_001 version of the pair.</p>
</li>
<li><p><strong>BAM processing, peak calling.</strong> It marks duplicates in bam files, and filter fragments by size.</p>
<pre><code>cd aligned<span class="hljs-selector-class">.aug10</span>
sbatch ./integrated<span class="hljs-selector-class">.step2</span><span class="hljs-selector-class">.sh</span> CR_BCL11A_W9_r1_S17_aligned_reads.bam
</code></pre></li>
<li><p><strong>Motif finding.</strong> CutRunTools uses MEME-chip for de novo motif finding on sequences surrounding the peak summits.</p>
<pre><code>cd ../macs2<span class="hljs-selector-class">.narrow</span><span class="hljs-selector-class">.aug18</span>
sbatch ./integrate<span class="hljs-selector-class">.motif</span><span class="hljs-selector-class">.find</span><span class="hljs-selector-class">.sh</span> CR_BCL11A_W9_r1_S17_aligned_reads_peaks.narrowPeak
</code></pre><p>By default, CutRunTools keeps duplicate fragments. If instead users wish to use deduplicate version,</p>
<pre><code>cd ../&lt;b&gt;macs2<span class="hljs-selector-class">.narrow</span><span class="hljs-selector-class">.aug18</span><span class="hljs-selector-class">.dedup</span>&lt;/b&gt;
sbatch ./integrate<span class="hljs-selector-class">.motif</span><span class="hljs-selector-class">.find</span><span class="hljs-selector-class">.sh</span> CR_BCL11A_W9_r1_S17_aligned_reads_peaks.narrowPeak
</code></pre></li>
<li><p><strong>Motif footprinting.</strong></p>
<pre><code>cd ../macs2<span class="hljs-selector-class">.narrow</span><span class="hljs-selector-class">.aug18</span>
sbatch ./integrate<span class="hljs-selector-class">.footprinting</span><span class="hljs-selector-class">.sh</span> CR_BCL11A_W9_r1_S17_aligned_reads_peaks.narrowPeak
</code></pre><p>Beautiful footprinting figures will be located in the directory <code>fimo.result</code>. Footprinting figures are created for every motif found by MEME-chip, but only the right motif (associated with TF) will have have a proper looking shape. Users can scan through individual motif&#39;s footprint.</p>
</li>
</ol>
</body>
</html>
